### PR TITLE
Adding callback to thingShadow.register

### PR DESCRIPTION
**Pull request by Torsph**
- Update the thing register function with a completion callback
- Add new test for the different states (ignoreDelta and persistentSubscribe)
- Updated README.md to reflect the new changes

Fixes issue #80 if I understood the aws and mqtt implementation correct. The new code waits for the mqtt lib to invoke the callback on _ignoreDeltas_ and _persistentSubscribe_ topics. When the mqtt lib has invoked both callbacks, will the register function invoke the register callback.

Maybe need for some more test cases, but please review and provide with feedback :)

### CODE REVIEW

File: thing/index.js [hunk 1]
```diff
@@ -477,15 +477,19 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
             discardStale: true,
             enableVersioning: true,
             qos: 0,
-            pending: false
+            pending: false,
+            pendingDeltaSubscribe: true,
+            pendingPersistentSubscribe: true
```
Comment (rongsaws) [inline thing/index.js]: I think it's better not to treat delta ad other shadow topic subscriptions separately. One flag should be enough to tell if the shadow is ready to take actions. 

In the same vein, callback should also be invoked only once to be deterministic and avoid user confusion.

File: thing/index.js [hunk 2]
```diff
@@ -515,7 +526,20 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
          //
          if (thingShadows[thingName].persistentSubscribe === true) {
             this._handleSubscriptions(thingName, ['update', 'get', 'delete'], ['accepted', 'rejected'],
-               'subscribe');
+               'subscribe', function () {
```
Comment (rongsaws) [inline thing/index.js]: subscribe callback should check granted subscriptions, see the underneath API here - https://github.com/mqttjs/MQTT.js#subscribe

If not all the topic subscriptions are granted, we could return an error in the register callback, and potentially fail following shadow operations.

File: thing/index.js [hunk 3]
```diff
@@ -477,15 +477,19 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
             discardStale: true,
             enableVersioning: true,
             qos: 0,
-            pending: false
+            pending: false,
+            pendingDeltaSubscribe: true,
```
Comment (rongsaws) [inline thing/index.js]: If this flag is pending, we should probably fail any shadow requests. We might be able to reuse the pending flag for the same purpose?

Comment (Torsph) [inline thing/index.js]: I guess we could. Will make a suggestion tomorrow or Monday.

File: thing/index.js (hunk 1 above)
Comment (Torsph) [inline thing/index.js]: Agree. I was thinking about that. I will make the needed changes :)

File: thing/index.js (hunk 2 above)
Comment (Torsph) [inline thing/index.js]: OK. We should also clean up the subscriptions and remove the shadow. Good to add more error handling.

File: test/mock/mockMQTTClient.js [hunk 4]
```diff
@@ -18,6 +18,7 @@ function mockMQTTClient( wrapper, options ) {
         this.subscribeQosValues.length = 0;
         this.publishQosValues = new Array;
         this.publishQosValues.length = 0;
+        this.triggerError = null;
```
Comment (Torsph) [inline test/mock/mockMQTTClient.js]: Maybe someone has a tip on how to force a error?

File: test/thing-unit-tests.js [hunk 5]
```diff
@@ -69,6 +69,80 @@ describe( "thing shadow class unit tests", function() {
...
+         caPath: 'test/data/root-CA.crt',
+         clientId: 'dummy-client-1',
+         region: 'us-east-1'
+      };
+
+      it("should trigger error when a subscription fails", function () {
+
+        // Need local stub to trigger error, so just copy-paste the code
+        var localMockMqtt;
+        var fakeConnect = function (options) {
```
Comment (Torsph) [inline test/thing-unit-tests.js]: Needed the possibility to force an error to get the test working. So reused the code from the beforeEach.

**Comment by Torsph** [comment]
Will clean up the branch if it's accepted. How do you like it? rebase into one commit or something else?

File: test/thing-unit-tests.js (hunk 5 above)
Comment (rongsaws) [inline test/thing-unit-tests.js]: This approach works, another similar option is to replace triggerError with a function, and by default it returns empty result, but you can stub it using Sinon so it returns subscription grant result to be passed to callback. You could stub it in your test case so you don't need to have the code repetition here.

File: thing/index.js [hunk 6]
```diff
@@ -526,6 +572,8 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
 
    this.unregister = function(thingName) {
       if (thingShadows.hasOwnProperty(thingName)) {
+         var topicsObject = [];
```
Comment (rongsaws) [inline thing/index.js]: Use topicObjects to be consistent with your prior definition; and perhaps topicSpecs or toipcDefinitions is more readable?

File: thing/index.js [hunk 7]
```diff
@@ -514,9 +549,20 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
...
-            this._handleSubscriptions(thingName, ['update', 'get', 'delete'], ['accepted', 'rejected'],
-               'subscribe');
+            topicsObject.push({
+               operations: ['update', 'get', 'delete'],
+               statii: ['accepted', 'rejected'],
+            });
+         }
+
+         if (topicsObject.length > 0) {
+            this._handleSubscriptions(thingName, topicsObject, 'subscribe', callback);
```
Comment (rongsaws) [inline thing/index.js]: From encapsulation point of view, pending should probably better be manipulated here, also you probably don't want to return granted etc in this register api, as whoever calls register() most likely won't care about what topics are accepted or not, what they do care is whether or not register() is successful. 

I would recommend to do something like this here

```
this._handleSubscriptions(thingName, topicsObject, 'subscribe', function(granted) {
        var res = "Not all subscriptions were granted";
        if (all topics are granted) {
               res = null;
               this.pending = false;  
        }
        if (!undefined(callback)) {
             callback(res);
        }
});
```


Also note _handleSubscriptions() does subscribe and unsubscribe, but for unsubscribe, there's no granted argument to the callback function.

**Comment by rongsaws** [comment]
Hey, thanks again for working on this. 

Your second commit looks good overall. If you don't mind, I could pull it in to our internal pipeline and have a reviewed by 2nd pair of eyes after making some minor adjustments? We could then have it released after that.

File: thing/index.js (hunk 6 above)
Comment (Torsph) [inline thing/index.js]: Agree

**Comment by Torsph** [comment]
Do you want me to fix the small issues before you pull it into the internal pipe?

**Comment by rongsaws** [comment]
Certainly if you are up for it:)

**Comment by Torsph** [comment]
Sure! Happy to help :)

File: thing/index.js (hunk 7 above)
Comment (Torsph) [inline thing/index.js]: Are you think about moving the logic for checking if all topics are granted into the callback? How are we going to handle all the other places that uses the _handleSubscriptions to sub?

Comment (rongsaws) [inline thing/index.js]: Checking granted code can be done in _handleSubscriptions() so users of it do need to check the ugly code 128 and so on. What I think it should pass to the callback is probably something like this:

callback(res, failedTopics) 

For successful subscriptions, both arguments would be null, for failed one, failed topics can be returned. For unsubscribe requests, failedTopics will always be empty.

File: test/thing-unit-tests.js [hunk 8]
```diff
@@ -69,6 +69,70 @@ describe( "thing shadow class unit tests", function() {
...
+         keyPath: 'test/data/private.pem.key',
+         certPath: 'test/data/certificate.pem.crt',
+         caPath: 'test/data/root-CA.crt',
+         clientId: 'dummy-client-1',
+         region: 'us-east-1'
+      };
+
+      it("should trigger error when a subscription fails", function () {
+
+        var stubTriggerError = sinon.stub(mockMQTTClient, 'triggerError', function(){return true;});
```
Comment (Torsph) [inline test/thing-unit-tests.js]: Haven't used sinon that much before, so don't know if I did it the correct way :)

**Comment by Torsph** [comment]
- Changed the named to topicSpecs
- Moved the pending assignment
- The callback returns the error message and the topics that failed.
- Updated the test with sinon stub. (Haven't used sinon that much so it may be wrong, but it works :))

Hope that will do the work. Please just do all the changes needed in your internal review :)

**Comment by Torsph** [comment]
Hi!

Do you have any update this PR? :)

**Comment by jimisataws** [comment]
Hi Torsph,

Thanks for your contributions. These changes are included in our next release cycle and currently going through the internal pipeline.

Thanks
Jimi

**Comment by liuszeng** [comment]
Addressed in release [v1.0.13](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.13).

**Pull request closed without merging**